cw/util/trace_library.py

```diff
@@ -0,0 +1,231 @@
+                      autoincrement=True),
+            db.Column("wave", db.LargeBinary),
+            db.Column("plaintext", db.LargeBinary),
+            db.Column("ciphertext", db.LargeBinary),
+            db.Column("key", db.LargeBinary),
+            db.Column("x_pos", db.Integer),
+            db.Column("y_pos", db.Integer),
+        )
+        self.metadata_table = db.Table(
+            "metadata",
+            self.metadata,
+            db.Column("config", db.String),
+            db.Column("datetime", db.DateTime),
+            db.Column("bitstream_path", db.String),
+            db.Column("binary_path", db.String),
+            db.Column("bitstream", db.LargeBinary),
+            db.Column("binary", db.LargeBinary),
+            db.Column("offset", db.Integer),
+            db.Column("sample_rate", db.Integer),
+            db.Column("scope_gain", db.Float),
+            db.Column("trigger_level", db.Float),
+            db.Column("time_div", db.Float),
+            db.Column("ot_sca_commit", db.String),
+            db.Column("notes", db.String)
+        )
+        self.metadata.create_all(self.engine)
+        self.trace_mem = []
+        self.trace_mem_thr = trace_threshold
+        self.settingsDict = {}
+        self.wave_datatype = wave_datatype
+
+    def flush_to_disk(self):
+        """ Writes traces from memory into database.
+        """
+        if self.trace_mem:
+            query = db.insert(self.traces_table)
+            traces = []
+            for trace in self.trace_mem:
+                tr = asdict(trace)
+                del tr["trace_id"]
+                traces.append(tr)
+            self.session.execute(query, traces)
+            self.session.commit()
+            self.trace_mem = []
+
+    def write_to_buffer(self, trace):
+        """ Write traces into memory or into storage.
+
+        The trace first is written into memory. When reaching the trace
+        memory threshold, the traces are flushed into the database storage.
+
+        Args:
+            trace: The trace to write.
+        """
+        self.trace_mem.append(trace)
+        if (len(self.trace_mem) >= self.trace_mem_thr):
+            self.flush_to_disk()
+
+    def get_traces(self, start=0, end=0):
+        """ Get traces from database and stored into RAM.
+
+        Fetch traces from start to end from database storage into RAM.
+
+        Args:
+            start: fetch traces from trace index start
+            end: to trace index end. If no start and end is provided, all
+                 all traces are returned.
+        Returns:
+            The traces from the database.
+        """
+        self.flush_to_disk()
+
+        if end == 0:
+            query = db.select(self.traces_table)
+        else:
+            query = db.select(self.traces_table).where(
+                (self.traces_table.c.trace_id >= start) &
+                (self.traces_table.c.trace_id <= end))
+
+        ResultProxy = self.session.execute(query)
```

Comment:
The capitalisation in the name is a bit shouty :-) But I don't think we actually need a name at all!
```python
        for trace in self.session.execute(query).fetchall():
...
```

And I think we can do better than that too :-D (untested)
```python
        return [Trace(**trace._mapping)
                for trace in self.session.execute(query).fetchall()]
```